src/game_api/flags.hpp

```diff
@@ -329,14 +319,6 @@ const char* special_visibility_flags[]{
     "22: ",
     "23: ",
     "24: ",
-    "25: In back layer",
-    "26: ",
-    "27: ",
-    "28: ",
-    "29: ",
-    "30: ",
-    "31: ",
-    "32: ",
```

Comment:
The UI was using that...

Reply:
By "using" you mean there is essential code that checks this "flag" or just a display thing?

it's kind of pointless and i don't want to encourage people to check some flag when it's just a bool value that can be checked way easier than the `test_flag(state.special_visibility_flags, 25)`

Reply:
I mean the only reason these char arrays exist is to print labels in ui.cpp, and you just removed part of the array without fixing the ui code... Now the ui just prints values from basecamp_dialogue_win_flags instead.

I know it's wrong, but it's still a funny thing to commit without fixing the ui code too. Changing the ui to only show checkboxes for the 6 flags in the 3 8bit variables would be fine.

Reply:
That template sure makes it simpler, but you also made some of the flag lists freakishly long, cause the for loops used to be shorter than the array. Yeah I know lazy arrays, but they are all from a time when we didn't know anything about anything.

Also there are a ton of unused flags between all the lists that ideally shouldn't be shown in the ui, cause things like presence flags and quest flags are also clearly multiple 8bit flags in a trench coat with a lot of unused flags in between. Unless those empty flags are used for something, but after a few years I think we would have documented them. I don't really care if some of these 32bit flags are actually multiple 8bit flags, as long as it's the whole 32 bits that is flags. It just makes the api more confusing to start slicing them all up now.

Reply:
The issue is, inconsistency in the flags. Some flags are empty (just number), other named "Unused" and some named "Unknown"

I can make the `render_flags` skip those if we really want to. Or add extra limit parameter to not show all of the fields

Also from quick look i think only `journal_flags` and `presence_flags` got longer

Reply:
We should probably

- make render_flags skip the whole checkbox if the name is empty
- remove numbers from flag names but
- add bool argument to print the numbers anyway, when knowing the flag number seems relevant
- change the few actually used but empty names to ???, unknown or best guess
- empty out the definitely unused names
- (add still missing flag enums to api, so flags.hpp is not essential for documentation)

Reply:
```c++
template <std::size_t SIZE, typename T>
void render_flags(const std::array<const char*, SIZE> names_array, T* flag_field, bool show_number = true)
{
    for (int idx{0}; idx < SIZE && idx < sizeof(T) * 8; ++idx)
    {
        // just simplified version of CheckboxFlagsT

        T value = (T)std::pow(2, idx);
        bool on = (*flag_field & value) == value;
        if (names_array[idx][0] != '\0' && ImGui::Checkbox(show_number ? fmt::format("{}: {}", idx + 1, names_array[idx]).c_str() : names_array[idx], &on))
        {
            *flag_field ^= value;
        }
    }
}
```

and

```lua
std::array presence_flags{
    "Udjat eye",
    "Black market",
    "Vlad's castle/drill",
    "",
    "",
    "",
    "",
    "",
    "Moon challenge",
    "Star challenge",
    "Sun challenge",
};
```

turning into

![image](https://github.com/spelunky-fyi/overlunky/assets/4947943/3d719931-3016-4e76-a4d7-816a56fc662b)

Flag numbers should probably usually be shown, cause it's pretty common to at least talk about them using numbers.

Reply:
Few notes:
`basecamp_dialogue_win_flags` - not used in UI
`movable_entity->buttons` - not editable in UI, would be good feature, but it requires virtual function hook right?
`char_states` - don't have numbers, when i feel they should?
`pause_types` - have flag values, not flags numbers, should probably change to numbers for consistency right?

Reply:
also some stuff is named `_flags` then it's not really a flag, just an array or something

Reply:
- basecamp_dialogue_win_flags: I guess someone just documented it there, but the whole dialogue thing seemed pretty useless or boring for modding and we never made any api for it
- buttons: yeah and that would be annoying to do in ui. editing buttons for most monsters isn't even very useful, cause changing states is usually how they move and attack. it's really just a leftover from investigating how monsters even work and can we control them
- char_states is used next to an editable input with the number
- pause: maybe, but then again the PAUSE enum is also just masks instead of flag numbers
- _flags: lazy copy pasting